File: Data Structures & Algorithms/interleaving-string/submission-1.cpp
class Solution {
public:
    bool solve(string s1,string s2,string s3, int i,int j,auto& dp){
        if(i == s1.size() && j == s2.size())return true;
        if(dp[i][j] != -1)return dp[i][j];
        int k = i+j;
        bool ans = false;
        if(i < s1.size() && s1[i] == s3[k]){
            ans = solve(s1,s2,s3,i+1,j,dp);
        }
        if(j < s2.size() && s2[j] == s3[k]){
            ans = ans || solve(s1,s2,s3,i,j+1,dp);
        }
        return dp[i][j] = ans;
    }
    bool isInterleave(string s1, string s2, string s3) {
        int l1 = s1.size() , l2 = s2.size();
        if(l1 + l2 != s3.size())return false;
        vector<vector<int>> dp(l1+1,vector<int>(l2+1,-1));
        return solve(s1,s2,s3,0,0,dp);
    }
};

File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-1.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    int ans;
    public Solution(){
        ans = 0;
    }
    public void solve(TreeNode root,int maxi){
        if(root == null)return;
        if(root.val >= maxi)ans++;
        maxi = Math.max(root.val,maxi);
        solve(root.left,maxi); 
        solve(root.right,maxi); 
    }
    public int goodNodes(TreeNode root) {
        solve(root,root.val);
        return ans;
    }
}
